guice/src/main/java/com/flipkart/gjex/grpc/interceptor/FilterInterceptor.java

```diff
@@ -169,9 +199,23 @@ public void onCancel() {
                 Context previous = attachContext(contextWithHeaders);   // attaching headers to gRPC context
                 try {
                     super.onCancel();
-                } catch (RuntimeException ex) {
-                    handleException(call, ex);
-                    grpcFilters.forEach(filter -> filter.doHandleException(ex));
+                    grpcFilters.forEach(filter -> filter.doHandleException(new Exception()));
+                    grpcFilters.forEach(Filter::doEndFilter);
+                } catch (Exception e) {
+                    handleException(call, grpcFilters, e);
+                } finally {
+                    detachContext(contextWithHeaders, previous);    // detach headers from gRPC context
+                }
+            }
+
+            @Override
+            public void onComplete() {
+                Context previous = attachContext(contextWithHeaders);   // attaching headers to gRPC context
+                try {
+                    super.onComplete();
+                    grpcFilters.forEach(Filter::doEndFilter);
+                } catch (Exception e) {
+                    handleException(call, grpcFilters, e);
```

Comment:
we shouldn't be calling handleException here. HandleException is for upstream. For Filter thrown error that needs to be returned without giving it back to the same filter.